File: 113-path-sum-ii/path-sum-ii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function(root, targetSum) {
    let result = [];
    dfs(root, targetSum, [], result);
    return result;
};

function dfs(root, targetSum, path, result) {
    if(root === null) return;
    path.push(root.val)
    targetSum-=root.val;

    if(root.left === null && root.right === null && targetSum === 0){
        result.push([...path])
    }
    dfs(root.left, targetSum, path, result);
    dfs(root.right, targetSum, path, result);
    path.pop();

}

File: 113-path-sum-ii/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        result = []
        self.Helper(root, targetSum, [], result)
        return result


    def Helper(self, root, targetSum,path, result):
        if root == None:
            return []
        path.append(root.val)
        targetSum -= root.val
        if root.left == None and root.right == None and targetSum == 0:
            result.append(path.copy())
        self.Helper(root.left, targetSum,path, result) 
        self.Helper(root.right, targetSum,path, result)
        path.pop()
        